### ISSUE TITLE
Datamapper SQL syntax error: argument passed to method computed as single quote instead of value

### ISSUE THREAD

**Issue by vladCovaliov** [datamapper/dm-core#291]
Hi everyone,

When I'm trying to execute someting like

`Table.first(username: username)` an SQL syntax error is raised.

I am using ruby 2.1.1 with DataMapper 1.2.0 on Mac OS X.

Gemfile

 ```
data_mapper (1.2.0)
      dm-aggregates (~> 1.2.0)
      dm-constraints (~> 1.2.0)
      dm-core (~> 1.2.0)
      dm-migrations (~> 1.2.0)
      dm-serializer (~> 1.2.0)
      dm-timestamps (~> 1.2.0)
      dm-transactions (~> 1.2.0)
      dm-types (~> 1.2.0)
      dm-validations (~> 1.2.0)
data_objects (0.10.14)
      addressable (~> 2.1)
    database_cleaner (1.3.0)
    debugger-linecache (1.2.0)
diff-lcs (1.2.5)
dm-aggregates (1.2.0)
  dm-core (~> 1.2.0)
dm-ar-finders (1.2.0)
  dm-core (~> 1.2.0)
dm-constraints (1.2.0)
  dm-core (~> 1.2.0)
dm-core (1.2.1)
  addressable (~> 2.3)
dm-do-adapter (1.2.0)
  data_objects (~> 0.10.6)
  dm-core (~> 1.2.0)
dm-migrations (1.2.0)
  dm-core (~> 1.2.0)
dm-mysql-adapter (1.2.0)
  dm-do-adapter (~> 1.2.0)
  do_mysql (~> 0.10.6)
dm-serializer (1.2.2)
  dm-core (~> 1.2.0)
  fastercsv (~> 1.5)
  json (~> 1.6)
  json_pure (~> 1.6)
  multi_json (~> 1.0)
dm-timestamps (1.2.0)
  dm-core (~> 1.2.0)
dm-transactions (1.2.0)
  dm-core (~> 1.2.0)
dm-types (1.2.2)
  bcrypt-ruby (~> 3.0)
  dm-core (~> 1.2.0)
  fastercsv (~> 1.5)
  json (~> 1.6)
  multi_json (~> 1.0)
  stringex (~> 1.4)
  uuidtools (~> 2.1)
dm-validations (1.2.0)
  dm-core (~> 1.2.0)
do_mysql (0.10.14)
  data_objects (= 0.10.14)
```

```
ERROR: Exception DataObjects::SQLError: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '') ORDER BY `id` LIMIT 1' at line 1 - stopping with code internal_server_error
ERROR: /Users/USERNAME/.rvm/gems/ruby-2.1.1@PROJECTNAME/gems/dm-do-adapter-1.2.0/lib/dm-do-adapter/adapter.rb:147:in `execute_reader'
/Users/USERNAME/.rvm/gems/ruby-2.1.1@PROJECTNAME/gems/dm-do-adapter-1.2.0/lib/dm-do-adapter/adapter.rb:147:in `block in read'
/Users/USERNAME/.rvm/gems/ruby-2.1.1@PROJECTNAME/gems/dm-do-adapter-1.2.0/lib/dm-do-adapter/adapter.rb:276:in `with_connection'
/Users/USERNAME/.rvm/gems/ruby-2.1.1@PROJECTNAME/gems/dm-do-adapter-1.2.0/lib/dm-do-adapter/adapter.rb:141:in `read'
/Users/USERNAME/.rvm/gems/ruby-2.1.1@PROJECTNAME/gems/dm-core-1.2.1/lib/dm-core/repository.rb:162:in `read'
/Users/USERNAME/.rvm/gems/ruby-2.1.1@PROJECTNAME/gems/dm-core-1.2.1/lib/dm-core/model.rb:378:in `first'
```

My SQL logs says:

```
You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to
 use near '') ORDER BY `id` LIMIT 1' at line 1 (code: 1064, sql state: 42000, 
 query: SELECT `a`, `b`, `c`, `q`, `d`, `e`, `f`, `g` FROM `table` WHERE (`something` = 3 AND `username` = ') ORDER BY `id` LIMIT 1, uri: MYURI
```

Please note the **username = '** part.

This is how my model looks:

```
class Table
    include DataMapper::Resource

    property    :id,                Serial,         key: true, max: BIGINT
    property    :username,          String,         required: true, unique_index: :strategy_username, _: 'Unique user id'
end
```

Important: Multiple apps are connected to the same SQL instance, but only one PC has this connection problem. Other queries are working from this PC but not this one with the username. IMO it's something related to gems or local environment settings, definitely not a code error. I tried reinstalling the gems , rvm and connecting to a similar SQL database but with no results.

Any help is really appreciated. Thanks!

**Comment by jay-wallaby** [comment]
Did anyone figure this out?  I am having the same issue running Ruby 1.9.3 and dm-core 1.2.1 on Mac OSX.

**Comment by tpitale** [comment]
No. It's possibly dependent on the MySQL version, oras bug that's already been fixed, or some other gem in the project.

You'd need to try against master. And also try recreating the problem with a simplified, new code.

Can you reproduce with just dm-core?

**Comment by jay-wallaby** [comment]
Sorry for the delay in response.  It's not the MySQL version; I've tried it with a 5.5 and a 5.7 DB and gotten the same error each time.  I'm working on a large existing code base, so it would take some time to put together a smaller example of the issue.  For the record, here are all of the DataMapper-related gems in the project:

`dm-adjust (1.2.0)
dm-aggregates (1.2.0)
dm-constraints (1.2.0)
dm-core (1.2.1)
dm-do-adapter (1.2.0)
dm-migrations (1.2.0)
dm-mysql-adapter (1.2.0)
dm-serializer (1.2.2)
dm-timestamps (1.2.0)
dm-validations (1.2.0)`

Other people working on the same project do not get this error, so I don't think it's one of the other gems.

I will let you know if I find the time to put together a smaller example, but for now my workaround is to develop in a virtual environment.  Sorry I couldn't be more helpful.

**Comment by tpitale** [comment]
So, it is something specific to your environment. If you can reproduce, we can take a look. Until then, I'm going to close this.

**Comment by jay-wallaby** [comment]
Just as a follow-up, I was able to fix this on my machine, and thought I would mention it here in case anybody found this page from a search engine like I did.  I tried uninstalling and reinstalling the entire project, which did not work.  I then uninstalled the project, uninstalled MySQL (which had been installed using Homebrew), reinstalled MySQL from a .dmg file, and reinstalled the project.  To get the do_mysql 0.10.13 gem to install I had to add `/usr/local/mysql/bin` to my PATH, but otherwise the installation went the same as before.  This time I did not get the syntax error when running MySQL queries.

So it seems to me that it's either a weird interaction with a Homebrew install, or it was a PATH issue, or I suppose a combination of the two things.  It might also have been more of a do_mysql gem problem caused by one of those things.
